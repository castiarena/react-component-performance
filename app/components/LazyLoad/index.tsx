import React, {
    FC, useState, useEffect, Suspense, lazy,
} from 'react';
import Loading from '../Loading';
import { Phar } from '../Typography';

interface LazyLoadProps {
    importComponent: () => any;
}

const LazyLoading = () => (
    <Loading>
        <Phar align="center">loading...</Phar>
    </Loading>
);

const LazyLoad: FC<LazyLoadProps> = ({ importComponent }) => {
    const [isFront, setIsFront] = useState(false);
    const LazyComponent = lazy(importComponent);

    useEffect(() => {
        process.nextTick(() => setIsFront(() => !!globalThis.window));
    }, [isFront]);

    if (!isFront) {
        return <LazyLoading />;
    }

    return (
        <Suspense fallback={<LazyLoading />}>
            <LazyComponent />
        </Suspense>
    );
};

export default LazyLoad;
