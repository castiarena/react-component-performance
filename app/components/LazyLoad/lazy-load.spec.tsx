import React from 'react';
import { ThemeProvider } from 'styled-components';
import 'jest-styled-components';
import { render } from '@testing-library/react';
import LazyLoad from '.';
import customTheme from '../../theme';

describe('Lazy load test suite', () => {
    it('should render as front', () => {
        const { asFragment, rerender } = render(
            <ThemeProvider theme={customTheme}>
                <LazyLoad importComponent={() => () => <div />} />
            </ThemeProvider>,
        );

        rerender(
            <ThemeProvider theme={customTheme}>
                <LazyLoad importComponent={() => () => <div />} />
            </ThemeProvider>,
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
