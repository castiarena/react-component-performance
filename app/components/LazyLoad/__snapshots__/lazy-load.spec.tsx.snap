// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lazy load test suite should render as front 1`] = `
<DocumentFragment>
  .c0 {
  padding: 26px;
}

.c1 {
  font-size: 1em;
  margin: 12px 0;
  line-height: 1.3em;
  text-align: center;
  color: #0A0A33;
  font-weight: normal;
}

<div
    class="c0"
  >
    <p
      class="c1"
    >
      loading...
    </p>
  </div>
</DocumentFragment>
`;
